core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java

```diff
@@ -709,78 +862,116 @@ public void writeToAllNonTxStores(MarshallableEntry marshalledEntry, int segment
    }
 
    @Override
-   public void writeBatchToAllNonTxStores(Iterable<MarshallableEntry> entries,
+   public CompletionStage<Void> writeToAllNonTxStores(MarshallableEntry marshalledEntry, int segment,
+         Predicate<? super StoreConfiguration> predicate, long flags) {
+      assert !Thread.currentThread().getName().startsWith("persistence") : "Thread name is: " + Thread.currentThread().getName();
+      int traceId = getNextTraceNumber("Writing to all stores for id %d");
+      return continueOnCPUExecutor(CompletableFuture.runAsync(() ->
+            writeToAllNonTxStoresSync(marshalledEntry, segment, predicate, flags, traceId), persistenceExecutor), traceId);
+   }
+
+   @Override
+   public CompletionStage<Void> writeBatchToAllNonTxStores(Iterable<MarshallableEntry> entries,
          Predicate<? super StoreConfiguration> predicate, long flags) {
       if (!entries.iterator().hasNext())
-         return;
+         return CompletableFutures.completedNull();
 
-      storesMutex.readLock().lock();
-      try {
-         checkStoreAvailability();
-         //noinspection unchecked
-         CompletableFuture batchFutures = CompletableFuture.allOf(
-               nonTxWriters.stream()
-               .filter(writer -> !(writer instanceof FlagAffectedStore) || FlagAffectedStore.class.cast(writer).shouldWrite(flags))
-               .filter(writer -> predicate.test(getStoreConfig(writer)))
-               .map(writer -> writer.bulkUpdate(Flowable.fromIterable(entries)))
-               .map(CompletionStage::toCompletableFuture)
-               .toArray(CompletableFuture[]::new)
-         );
+      assert !Thread.currentThread().getName().startsWith("persistence") : "Thread name is: " + Thread.currentThread().getName();
+
+      int id = getNextTraceNumber("Submitting persistence async operation of id %d to write a batch");
+
+      return continueOnCPUExecutor(CompletableFuture.runAsync(() -> {
+         if (trace) {
+            log.tracef("Continuing write batch for id %d", id);
+         }
+         storesMutex.readLock().lock();
          try {
-            batchFutures.get();
-         } catch (ExecutionException e) {
-            throw new PersistenceException(e.getCause());
-         } catch (InterruptedException e) {
-            Thread.currentThread().interrupt();
-            throw new PersistenceException(e);
+            checkStoreAvailability();
+            //noinspection unchecked
+            CompletableFuture batchFutures = CompletableFuture.allOf(
+                  nonTxWriters.stream()
+                  .filter(writer -> !(writer instanceof FlagAffectedStore) || FlagAffectedStore.class.cast(writer).shouldWrite(flags))
+                  .filter(writer -> predicate.test(getStoreConfig(writer)))
+                        // TOOD: the bulkUpdate has good API but implementations are still blocking
```

Comment:
Separate PR?

Reply:
No, I should really treat it as if the implementations are non blocking - I will fix the code to do so.